builtin.c

```diff
@@ -273,33 +272,26 @@ XS(XS_builtin_trim)
         }
     }
 
-    dest = TARG;
+    SV_CHECK_THINKFIRST(TARG);
+    SvUPGRADE(TARG, SVt_PV);
+    SvGROW(TARG, len + 1);
 
-    if (SvPOK(dest) && (dest == source)) {
-        sv_chop(dest, (const char *)start);
-        SvCUR_set(dest, len);
-    }
-    else {
-        SvUPGRADE(dest, SVt_PV);
-        SvGROW(dest, len + 1);
-
-        Copy(start, SvPVX(dest), len, U8);
-        SvPVX(dest)[len] = '\0';
-        SvPOK_on(dest);
-        SvCUR_set(dest, len);
+    Copy(start, SvPVX(TARG), len, U8);
```

Comment:
Is `start` always still valid after the THINKFIRST? Is it possible that could copy out a CoW string to a new buffer?

Oh but then I suppose `start` will still point at the original untouched buffer so it's probably all fine.

Reply:
> Oh but then I suppose `start` will still point at the original untouched buffer so it's probably all fine.

That's what I expect.

Reply:
About https://github.com/Perl/perl5/commit/4539fb8c81b73cb8057859ccab110502c1db28b3 

Not saying the fix in blead/this ticket is wrong, but there might a better way to do it, or in general for XS "do it".

Note the old code removed in this commit used the *VERY* underused (core/cpan) SvOOK flag optimization. The new code does not have `sv_chop()` optimization to prevent double buffering and maybe a trip through `re/malloc()`. 

The 5.25 newish `sv_set_undef()` to get a no-leak SvOK_off() then sv_chop(), or SvCUR() lowering, which is changing 1 or 2 integers only, vs Move()/memcpy(), is the better implementation for white space trimming a string. Perl has the API to do it efficiently (string manipulation), SvOOK needs to be used more often by the community.

Reply:
> Note the old code removed in this commit used the VERY underused (core/cpan) SvOOK flag optimization. The new code does not have sv_chop() optimization to prevent double buffering and maybe a trip through re/malloc().

The sv_chop() branch was removed because the input and output SVs are never the same SV.

> The 5.25 newish sv_set_undef() ...

I have a [clang-tidy check for that](https://github.com/tonycoz/perl-check#perl-check) (though I believe loadable clang-tidy checks don't work on Windows).

Reply:
Can you discreetly slip that magic potion into EU::ParseXS when nobody is looking??!?!  :pleading_face: